data/srv/epages/eproot/Perl/man/man3/lib.3
.\" Automatically generated by Pod::Man 2.23 (Pod::Simple 3.14)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "lib 3"
.TH lib 3 "2011-04-13" "perl v5.12.3" "Perl Programmers Reference Guide"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
lib \- manipulate @INC at compile time
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&    use lib LIST;
\&
\&    no lib LIST;
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This is a small simple module which simplifies the manipulation of \f(CW@INC\fR
at compile time.
.PP
It is typically used to add extra directories to perl's search path so
that later \f(CW\*(C`use\*(C'\fR or \f(CW\*(C`require\*(C'\fR statements will find modules which are
not located on perl's default search path.
.ie n .SS "Adding directories to @INC"
.el .SS "Adding directories to \f(CW@INC\fP"
.IX Subsection "Adding directories to @INC"
The parameters to \f(CW\*(C`use lib\*(C'\fR are added to the start of the perl search
path. Saying
.PP
.Vb 1
\&    use lib LIST;
.Ve
.PP
is \fIalmost\fR the same as saying
.PP
.Vb 1
\&    BEGIN { unshift(@INC, LIST) }
.Ve
.PP
For each directory in \s-1LIST\s0 (called \f(CW$dir\fR here) the lib module also
checks to see if a directory called \f(CW$dir\fR/$archname/auto exists.
If so the \f(CW$dir\fR/$archname directory is assumed to be a corresponding
architecture specific directory and is added to \f(CW@INC\fR in front of \f(CW$dir\fR.
lib.pm also checks if directories called \f(CW$dir\fR/$version and \f(CW$dir\fR/$version/$archname
exist and adds these directories to \f(CW@INC\fR.
.PP
The current value of \f(CW$archname\fR can be found with this command:
.PP
.Vb 1
\&    perl \-V:archname
.Ve
.PP
The corresponding command to get the current value of \f(CW$version\fR is:
.PP
.Vb 1
\&    perl \-V:version
.Ve
.PP
To avoid memory leaks, all trailing duplicate entries in \f(CW@INC\fR are
removed.
.ie n .SS "Deleting directories from @INC"
.el .SS "Deleting directories from \f(CW@INC\fP"
.IX Subsection "Deleting directories from @INC"
You should normally only add directories to \f(CW@INC\fR.  If you need to
delete directories from \f(CW@INC\fR take care to only delete those which you
added yourself or which you are certain are not needed by other modules
in your script.  Other modules may have added directories which they
need for correct operation.
.PP
The \f(CW\*(C`no lib\*(C'\fR statement deletes all instances of each named directory
from \f(CW@INC\fR.
.PP
For each directory in \s-1LIST\s0 (called \f(CW$dir\fR here) the lib module also
checks to see if a directory called \f(CW$dir\fR/$archname/auto exists.
If so the \f(CW$dir\fR/$archname directory is assumed to be a corresponding
architecture specific directory and is also deleted from \f(CW@INC\fR.
.ie n .SS "Restoring original @INC"
.el .SS "Restoring original \f(CW@INC\fP"
.IX Subsection "Restoring original @INC"
When the lib module is first loaded it records the current value of \f(CW@INC\fR
in an array \f(CW@lib::ORIG_INC\fR. To restore \f(CW@INC\fR to that value you
can say
.PP
.Vb 1
\&    @INC = @lib::ORIG_INC;
.Ve
.SH "CAVEATS"
.IX Header "CAVEATS"
In order to keep lib.pm small and simple, it only works with Unix
filepaths.  This doesn't mean it only works on Unix, but non-Unix
users must first translate their file paths to Unix conventions.
.PP
.Vb 3
\&    # VMS users wanting to put [.stuff.moo] into 
\&    # their @INC would write
\&    use lib \*(Aqstuff/moo\*(Aq;
.Ve
.SH "NOTES"
.IX Header "NOTES"
In the future, this module will likely use File::Spec for determining
paths, as it does now for Mac \s-1OS\s0 (where Unix-style or Mac-style paths
work, and Unix-style paths are converted properly to Mac-style paths
before being added to \f(CW@INC\fR).
.PP
If you try to add a file to \f(CW@INC\fR as follows:
.PP
.Vb 1
\&  use lib \*(Aqthis_is_a_file.txt\*(Aq;
.Ve
.PP
\&\f(CW\*(C`lib\*(C'\fR will warn about this. The sole exceptions are files with the
\&\f(CW\*(C`.par\*(C'\fR extension which are intended to be used as libraries.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
FindBin \- optional module which deals with paths relative to the source file.
.PP
\&\s-1PAR\s0 \- optional module which can treat \f(CW\*(C`.par\*(C'\fR files as Perl libraries.
.SH "AUTHOR"
.IX Header "AUTHOR"
Tim Bunce, 2nd June 1995.
.PP
\&\f(CW\*(C`lib\*(C'\fR is maintained by the perl5\-porters. Please direct
any questions to the canonical mailing list. Anything that
is applicable to the \s-1CPAN\s0 release can be sent to its maintainer,
though.
.PP
Maintainer: The Perl5\-Porters <perl5\-porters@perl.org>
.PP
Maintainer of the \s-1CPAN\s0 release: Steffen Mueller <smueller@cpan.org>
.SH "COPYRIGHT AND LICENSE"
.IX Header "COPYRIGHT AND LICENSE"
This package has been part of the perl core since perl 5.001.
It has been released separately to \s-1CPAN\s0 so older installations
can benefit from bug fixes.
.PP
This package has the same copyright and license as the perl core.


data/srv/epages/eproot/Perl/lib/linux/lib.pm
package lib;

# THIS FILE IS AUTOMATICALLY GENERATED FROM lib_pm.PL.
# ANY CHANGES TO THIS FILE WILL BE OVERWRITTEN BY THE NEXT PERL BUILD.

use Config;

use strict;

my $archname         = $Config{archname};
my $version          = $Config{version};
my @inc_version_list = reverse split / /, $Config{inc_version_list};


our @ORIG_INC = @INC;   # take a handy copy of 'original' value
our $VERSION = '0.62';
my $Is_MacOS = $^O eq 'MacOS';
my $Mac_FS;
if ($Is_MacOS) {
        require File::Spec;
        $Mac_FS = eval { require Mac::FileSpec::Unixish };
}

sub import {
    shift;

    my %names;
    foreach (reverse @_) {
        my $path = $_;          # we'll be modifying it, so break the alias
        if ($path eq '') {
            require Carp;
            Carp::carp("Empty compile time value given to use lib");
        }

        $path = _nativize($path);

        if ($path !~ /\.par$/i && -e $path && ! -d _) {
            require Carp;
            Carp::carp("Parameter to use lib must be directory, not file");
        }
        unshift(@INC, $path);
        # Add any previous version directories we found at configure time
        foreach my $incver (@inc_version_list)
        {
            my $dir = $Is_MacOS
                ? File::Spec->catdir( $path, $incver )
                : "$path/$incver";
            unshift(@INC, $dir) if -d $dir;
        }
        # Put a corresponding archlib directory in front of $path if it
        # looks like $path has an archlib directory below it.
        my($arch_auto_dir, $arch_dir, $version_dir, $version_arch_dir)
            = _get_dirs($path);
        unshift(@INC, $arch_dir)         if -d $arch_auto_dir;
        unshift(@INC, $version_dir)      if -d $version_dir;
        unshift(@INC, $version_arch_dir) if -d $version_arch_dir;
    }

    # remove trailing duplicates
    @INC = grep { ++$names{$_} == 1 } @INC;
    return;
}


sub unimport {
    shift;

    my %names;
    foreach (@_) {
        my $path = _nativize($_);

        my($arch_auto_dir, $arch_dir, $version_dir, $version_arch_dir)
            = _get_dirs($path);
        ++$names{$path};
        ++$names{$arch_dir}         if -d $arch_auto_dir;
        ++$names{$version_dir}      if -d $version_dir;
        ++$names{$version_arch_dir} if -d $version_arch_dir;
    }

    # Remove ALL instances of each named directory.
    @INC = grep { !exists $names{$_} } @INC;
    return;
}

sub _get_dirs {
    my($dir) = @_;
    my($arch_auto_dir, $arch_dir, $version_dir, $version_arch_dir);

    # we could use this for all platforms in the future, but leave it
    # Mac-only for now, until there is more time for testing it.
    if ($Is_MacOS) {
        $arch_auto_dir    = File::Spec->catdir( $dir, $archname, 'auto' );
        $arch_dir         = File::Spec->catdir( $dir, $archname, );
        $version_dir      = File::Spec->catdir( $dir, $version );
        $version_arch_dir = File::Spec->catdir( $dir, $version, $archname );
    } else {
        $arch_auto_dir    = "$dir/$archname/auto";
        $arch_dir         = "$dir/$archname";
        $version_dir      = "$dir/$version";
        $version_arch_dir = "$dir/$version/$archname";
    }
    return($arch_auto_dir, $arch_dir, $version_dir, $version_arch_dir);
}

sub _nativize {
    my($dir) = @_;

    if ($Is_MacOS && $Mac_FS && ! -d $dir) {
        $dir = Mac::FileSpec::Unixish::nativize($dir);
        $dir .= ":" unless $dir =~ /:$/;
    }

    return $dir;
}

1;
__END__

=head1 NAME

lib - manipulate @INC at compile time

=head1 SYNOPSIS

    use lib LIST;

    no lib LIST;

=head1 DESCRIPTION

This is a small simple module which simplifies the manipulation of @INC
at compile time.

It is typically used to add extra directories to perl's search path so
that later C<use> or C<require> statements will find modules which are
not located on perl's default search path.

=head2 Adding directories to @INC

The parameters to C<use lib> are added to the start of the perl search
path. Saying

    use lib LIST;

is I<almost> the same as saying

    BEGIN { unshift(@INC, LIST) }

For each directory in LIST (called $dir here) the lib module also
checks to see if a directory called $dir/$archname/auto exists.
If so the $dir/$archname directory is assumed to be a corresponding
architecture specific directory and is added to @INC in front of $dir.
lib.pm also checks if directories called $dir/$version and $dir/$version/$archname
exist and adds these directories to @INC.

The current value of C<$archname> can be found with this command:

    perl -V:archname

The corresponding command to get the current value of C<$version> is:

    perl -V:version

To avoid memory leaks, all trailing duplicate entries in @INC are
removed.

=head2 Deleting directories from @INC

You should normally only add directories to @INC.  If you need to
delete directories from @INC take care to only delete those which you
added yourself or which you are certain are not needed by other modules
in your script.  Other modules may have added directories which they
need for correct operation.

The C<no lib> statement deletes all instances of each named directory
from @INC.

For each directory in LIST (called $dir here) the lib module also
checks to see if a directory called $dir/$archname/auto exists.
If so the $dir/$archname directory is assumed to be a corresponding
architecture specific directory and is also deleted from @INC.

=head2 Restoring original @INC

When the lib module is first loaded it records the current value of @INC
in an array C<@lib::ORIG_INC>. To restore @INC to that value you
can say

    @INC = @lib::ORIG_INC;

=head1 CAVEATS

In order to keep lib.pm small and simple, it only works with Unix
filepaths.  This doesn't mean it only works on Unix, but non-Unix
users must first translate their file paths to Unix conventions.

    # VMS users wanting to put [.stuff.moo] into
    # their @INC would write
    use lib 'stuff/moo';

=head1 NOTES

In the future, this module will likely use File::Spec for determining
paths, as it does now for Mac OS (where Unix-style or Mac-style paths
work, and Unix-style paths are converted properly to Mac-style paths
before being added to @INC).

If you try to add a file to @INC as follows:

  use lib 'this_is_a_file.txt';

C<lib> will warn about this. The sole exceptions are files with the
C<.par> extension which are intended to be used as libraries.

=head1 SEE ALSO

FindBin - optional module which deals with paths relative to the source file.

PAR - optional module which can treat C<.par> files as Perl libraries.

=head1 AUTHOR

Tim Bunce, 2nd June 1995.

C<lib> is maintained by the perl5-porters. Please direct
any questions to the canonical mailing list. Anything that
is applicable to the CPAN release can be sent to its maintainer,
though.

Maintainer: The Perl5-Porters <perl5-porters@perl.org>

Maintainer of the CPAN release: Steffen Mueller <smueller@cpan.org>

=head1 COPYRIGHT AND LICENSE

This package has been part of the perl core since perl 5.001.
It has been released separately to CPAN so older installations
can benefit from bug fixes.

This package has the same copyright and license as the perl core.

=cut
